constraint/pkg/client/drivers/k8scel/driver.go

```diff
@@ -33,166 +44,382 @@ package k8scel
-
-// type mockObjectInterfaces struct{}
-
-// type mockObjectCreater struct{}
-
-// func (m *mockObjectCreater) New(gvk schema.GroupVersionKind) (runtime.Object, error) {
-// 	return nil, errors.New("OBJECT CREATOR NOT IMPLEMENTED")
-// }
-
-// func (m *mockObjectInterfaces) GetObjectCreater() runtime.ObjectCreater { return &mockObjectCreater{} }
-
-// func (m *mockObjectInterfaces) GetObjectTyper() runtime.ObjectTyper { return nil }
-
-// func (m *mockObjectInterfaces) GetObjectDefaulter() runtime.ObjectDefaulter { return nil }
-
-// func (m *mockObjectInterfaces) GetObjectConvertor() runtime.ObjectConvertor { return nil }
-
-// func (m *mockObjectInterfaces) GetEquivalentResourceMapper() runtime.EquivalentResourceMapper {
-// 	return nil
-// }
+//   Other friction points are commented with the keyword FRICTION.
+
+const (
+	Name = "K8sValidation"
+
+	runTimeNS            = "runTimeNS"
+	runTimeNSDescription = "the number of nanoseconds it took to evaluate the constraint"
+)
+
+var _ drivers.Driver = &Driver{}
+
+type Driver struct {
+	mux            sync.RWMutex
+	validators     map[string]validatingadmissionpolicy.Validator
+	filterCompiler cel.FilterCompiler
+	gatherStats    bool
+}
+
+func (d *Driver) Name() string {
+	return Name
+}
+
+func (d *Driver) AddTemplate(ctx context.Context, ct *templates.ConstraintTemplate) error {
+	if len(ct.Spec.Targets) != 1 {
+		return errors.New("wrong number of targets defined, only 1 target allowed")
+	}
+	var rawCode map[string]interface{}
+	for _, code := range ct.Spec.Targets[0].Code {
+		if code.Engine != Name {
+			continue
+		}
+		objMap, ok := code.Source.Value.(map[string]interface{})
+		if !ok {
+			return errors.New("K8sValidation code malformed")
+		}
+		rawCode = objMap
+		break
+	}
+	if rawCode == nil {
+		return errors.New("K8sValidation code not defined")
+	}
+
+	validatorCode := &admissionv1alpha1.ValidatingAdmissionPolicy{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructuredWithValidation(rawCode, validatorCode, true); err != nil {
+		return err
+	}
+
+	// FRICTION: Note that compilation errors are possible, but we cannot introspect to see whether any
+	// occurred
+	// TODO can set this based on whether params is defined on a constraint
+	celVars := cel.OptionalVariableDeclarations{HasParams: true}
+	failurePolicy := convertv1alpha1FailurePolicyTypeTov1FailurePolicyType(validatorCode.Spec.FailurePolicy)
+	var matcher matchconditions.Matcher
+	matchConditions := validatorCode.Spec.MatchConditions
+	if len(matchConditions) > 0 {
+		matchExpressionAccessors := make([]cel.ExpressionAccessor, len(matchConditions))
+		for i := range matchConditions {
+			matchExpressionAccessors[i] = (*matchconditions.MatchCondition)(&matchConditions[i])
+		}
+		matcher = matchconditions.NewMatcher(d.filterCompiler.Compile(matchExpressionAccessors, celVars, 10000000), nil, failurePolicy, "validatingadmissionpolicy", validatorCode.Name)
```

Comment:
do we want to make `10000000` a const?

Reply:
what is that used for?

Reply:
That is the compute budget cap for executing a policy. Taking it from the CEL KEP's source code.

Reply:
Found the k8s constant, using it directly